onnxruntime/core/platform/windows/env.cc

```diff
@@ -202,643 +229,636 @@ class WindowsThread : public EnvThread {
+  // TODO: Understand the reason for failure by calling GetLastError().
+  // If it is due to the specified environment variable being found in the environment block,
+  // GetLastError() returns ERROR_ENVVAR_NOT_FOUND.
+  // For now, we assume that the environment variable is not found.
 
-    return std::string();
+  return std::string();
+}
+
+/*
+Read logical processor info from the map.
+{-1,-1} stands for failure.
+*/
+ProcessorInfo WindowsEnv::GetProcessorAffinityMask(int global_processor_id) const {
+  if (global_processor_info_map_.count(global_processor_id)) {
+    return global_processor_info_map_.at(global_processor_id);
+  } else {
+    return {-1, -1};
   }
+}
 
- private:
-  WindowsEnv() = default;
+WindowsEnv::WindowsEnv() {
+  InitializeCpuInfo();
+}
 
-  typedef VOID(WINAPI* FnGetSystemTimePreciseAsFileTime)(LPFILETIME);
-  WindowsTelemetry telemetry_provider_;
-};
-}  // namespace
+/*
+Discover all cores in a windows system.
+Note - every "id" here, given it be group id, core id, or logical processor id, starts from 0.
+*/
+void WindowsEnv::InitializeCpuInfo() {
+  DWORD returnLength = 0;
+  GetLogicalProcessorInformationEx(RelationProcessorCore, nullptr, &returnLength);
+  auto last_error = GetLastError();
+  if (last_error != ERROR_INSUFFICIENT_BUFFER) {
+    const auto error_code = GetLastError();
+    if (logging::LoggingManager::HasDefaultLogger()) {
+      LOGS_DEFAULT(ERROR) << "Failed to calculate byte size for saving cpu info on windows"
+                          << ", error code: " << error_code
+                          << ", error msg: " << std::system_category().message(error_code);
+    }
+    return;
+  }
 
-Env& Env::Default() {
-  return WindowsEnv::Instance();
+  std::unique_ptr<char[]> allocation = std::make_unique<char[]>(returnLength);
+  SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX* processorInfos = reinterpret_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*>(allocation.get());
+
+  if (!GetLogicalProcessorInformationEx(RelationProcessorCore, processorInfos, &returnLength)) {
+    const auto error_code = GetLastError();
+    if (logging::LoggingManager::HasDefaultLogger()) {
+      LOGS_DEFAULT(ERROR) << "Failed to fetch cpu info on windows"
+                          << ", error code: " << error_code
+                          << ", error msg: " << std::system_category().message(error_code);
+    }
+    return;
+  }
+
+  int core_id = 0;
+  int global_processor_id = 0;
+  const BYTE* iter = reinterpret_cast<const BYTE*>(processorInfos);
+  const BYTE* end = iter + returnLength;
+  std::stringstream log_stream;
+
+  while (iter < end) {
+    auto processor_info = reinterpret_cast<const SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*>(iter);
+    auto size = processor_info->Size;
+
+    //Discoverred a phyical core and it belongs exclusively to a single group
+    if (processor_info->Relationship == RelationProcessorCore &&
+        processor_info->Processor.GroupCount == 1) {
+      log_stream << std::endl
```

Comment:
> std::endl [](http://example.com/codeflow?start=19&length=10)

We do not need to flush to std::stringstream. A simple '\n' would be fine.

Reply:
Adding std::endl here to format output like:

```
2022-12-06 19:25:43.5940663 [V:onnxruntime:Default, env.cc:861 onnxruntime::WindowsEnv::InitializeCpuInfo]
core 1 consist of logical processors: 1 2
core 2 consist of logical processors: 3 4
core 3 consist of logical processors: 5 6
core 4 consist of logical processors: 7 8
```

otherwise it would be:

```
2022-12-06 19:25:43.5940663 [V:onnxruntime:Default, env.cc:861 onnxruntime::WindowsEnv::InitializeCpuInfo] core 1 consist of logical processors: 1 2
core 2 consist of logical processors: 3 4
core 3 consist of logical processors: 5 6
core 4 consist of logical processors: 7 8
```

Reply:
Prefer std::endl, because for windows it is "\r\n", while for linux it is "\n", for Mac it might "\r" or "\n" depending on the version ...